### ISSUE TITLE
Unable to Upgrade Cluster Configuration at All: Start-ServiceFabricClusterConfigurationUpgrade incorrectly identifies an incompatible authentication type change

### ISSUE THREAD

**Issue by JustinKaffenberger** [microsoft/service-fabric#327]
I've been learning how to stand up a cluster with X509 authentication, and encountered an issue with the `Start-ServiceFabricClusterConfigurationUpgrade` incorrectly identifying an incompatible authentication type change. The reality is, I am not changing the authentication type, but merely trying to update the `ApplicationPorts.PortRange`. I will go over my reproduction steps here:

1. Create a valid cluster following the template `.\Microsoft.Azure.ServiceFabric.WindowsServer.6.4.654.9590\ClusterConfig.X509.MultiMachine.json` that is provided with the on-premise install package.
2. Deploy a small Service Fabric application to validate that the install is stable and correct. In my case I am able to use a client certificate to access my remote cluster and deploy my application. The application successfully deploys.
3. I created a Powershell script to automate the process of pulling down the existing configuration, modifying the ports and the version, and then invoking the  `Start-ServiceFabricClusterConfigurationUpgrade` cmdlet.

```
# This code assumes we are executing out of the
# Microsoft.Azure.ServiceFabric.WindowsServer.6.4.654.9590 
# installation package folder

# Load the previous configuration from the cluster into an in-memory JSON object
# that we can modify.
$previousConfig = (Join-Path $env:TEMP ([string][System.Guid]::NewGuid() + ".json"))
Get-ServiceFabricClusterConfiguration | Out-File $previousConfig
$scratchJson = Get-Content $previousConfig | ConvertFrom-Json

# Modify the JSON as needed. In this case I am just modifying the end port.
$scratchJson.Properties.NodeTypes[0].ApplicationPorts.EndPort = 22048 # This is the end port I wanted to use
$scratchJson.ClusterConfigurationVersion = "1.0.1"

# Export the modified in-memory JSON object to a file
$newConfig = (Join-Path $env:TEMP ([string][System.Guid]::NewGuid() + ".json"))
$scratchJson | ConvertTo-Json -Depth 32 | Out-File $newConfig 

# Run the TestConfiguration cmdlet in the installer folder. I am seeing the following:
#
# Trace folder already exists. Traces will be written to existing trace folder: C:\Users\Administrator\Downloads\Microsoft.Azure.ServiceFabric.WindowsServer.6.4.654.9590\DeploymentTraces
# Running Best Practices Analyzer...
# Running Best Practices Upgrade Analyzer. Please make sure you are validating against the current cluster json config in your cluster. Cluster configuration in the system may change after you add or remove node which means the original cluster json config can't represent the cluster state anymore. You can get the latest configuration by Get-ServiceFabricClusterConfiguration in Powershell.
# Best Practices Analyzer completed successfully.
# LocalAdminPrivilege        : True
# IsJsonValid                : True
# IsCabValid                 :
# RequiredPortsOpen          : True
# RemoteRegistryAvailable    : True
# FirewallAvailable          : True
# RpcCheckPassed             : True
# NoDomainController         : True
# NoConflictingInstallations : True
# FabricInstallable          :
# DataDrivesAvailable        : True
# DrivesEnoughAvailableSpace : True
# IsAllOrNoneIOTDevice       : True
# DotNetExeInPath            : True
# Passed                     : True
#
.\TestConfiguration.ps1 -ClusterConfigFilePath $newConfig -OldClusterConfigFilePath $previousConfig -Verbose

Start-ServiceFabricClusterConfigurationUpgrade -ClusterConfigPath $newConfig
```

The result of this script is the following exception:

```
Start-ServiceFabricClusterConfigurationUpgrade : System.Runtime.InteropServices.COMException (-2147017627)
ValidationException: Authentication type cannot be changed from unsecured to Windows.
At line:1 char:1
+ Start-ServiceFabricClusterConfigurationUpgrade -ClusterConfigPath $ne ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (Microsoft.Servi...usterConnection:ClusterConnection) [Start-ServiceFa
   ...gurationUpgrade], FabricException
    + FullyQualifiedErrorId : StartClusterConfigurationUpgradeErrorId,Microsoft.ServiceFabric.Powershell.StartClusterC
   onfigurationUpgrade
```

At no point is my cluster unsecured, and the new config is not set to use Windows authentication, so I'm assuming this is a bug. Any help would be appreciated.

**Comment by ojaimi** [comment]
I also receive the above mentioned error in an intermittent way, sometimes I get it and sometimes I dont, below steps to reproduce it:

Connect to cluster securely using X509 Certificate 
`$connectArgs = @{  ConnectionEndpoint = 'XXXXXXXXXXXXXXXXXX:19000';  
                   X509Credential = $True;  
                   StoreLocation = "CurrentUser";  
                   StoreName = "My";  
                   ServerCommonName = "XXXXXXXXXXXXXXXXXX";  
                   FindType = 'FindByThumbprint';  
                   FindValue = "XXXXXXXXXXXXXXXXXX"; 
                   ServerCertThumbprint = "XXXXXXXXXXXXXXXXXX" }
Connect-ServiceFabricCluster @connectArgs`

Get current cluster configuration, modify needed lines and make sure you modify config file version
`Get-ServiceFabricClusterConfiguration -UseApiVersion -ApiVersion 10-2017 > ClusterConfigNew.json`

Test Modified file versus old one
`Test-ServiceFabricConfiguration  -FabricRuntimePackagePath '.\MicrosoftAzureServiceFabric.6.4.644.9590.cab' -ClusterConfigurationFilePath '.\ClusterConfigNew.json' -OldClusterConfigurationFilePath '.\ClusterConfigOld.json'`

> Running Best Practices Analyzer...
Running Best Practices Upgrade Analyze
your cluster. Cluster configuration in
er json config can't represent the clu
erConfiguration in Powershell.
Best Practices Analyzer completed succ


LocalAdminPrivilege        : True
IsJsonValid                : True
IsCabValid                 : True
RequiredPortsOpen          : True
RemoteRegistryAvailable    : True
FirewallAvailable          : True
RpcCheckPassed             : True
NoDomainController         : True
NoConflictingInstallations : True
FabricInstallable          :
DataDrivesAvailable        : True
DrivesEnoughAvailableSpace : True
IsAllOrNoneIOTDevice       : True
DotNetExeInPath            : True
Passed                     : True

`Start-ServiceFabricClusterConfigurationUpgrade -ClusterConfigPath .\ClusterConfigNew.json`

> Start-ServiceFabricClusterConfigurationUpgrade : System.Runtime.InteropServices.COMException (-2147017627)
ValidationException: A configuration upgrade with identical configuration version is not allowed.
At line:1 char:1
+ Start-ServiceFabricClusterConfigurationUpgrade -ClusterConfigPath .\C ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (Microsoft.Servi...usterConnection:ClusterConnection) [Start-ServiceFa
   ...gurationUpgrade], FabricException
    + FullyQualifiedErrorId : StartClusterConfigurationUpgradeErrorId,Microsoft.ServiceFabric.Powershell.StartClusterC
   onfigurationUpgrade

**Comment by andrewtyped** [comment]
I am also experiencing this with the same repro steps as @ojaimi . In my case, the only elements changed in  my cluster config were ClusterConfigurationVersion and FabricClusterAutoupgradeEnabled (from true to false). I don't feel comfortable sharing my full clusterConfig, but if there's a section I can provide let me know.

**Comment by JustinKaffenberger** [comment]
I updated the title to hopefully reflect the severity of this issue. We are completely unable to upgrade the cluster configuration due to this issue.

**Comment by JustinKaffenberger** [comment]
This is utter insanity that I can't upgrade my on-premise Service Fabric cluster that is using X509 authentication, and the issue hasn't even been commented on by anyone in over 3 months. How am I supposed to use Service Fabric in production if I can't perform basic cluster configuration upgrades when the time comes? I've also posted the issue [here](https://github.com/Azure/service-fabric-issues/issues/1512), thinking I posted in the wrong location, but still, no response!

**Comment by JustinKaffenberger** [comment]
I will close this for now because its clear that this repository is Linux focused (for now), and keep the conversation going [here](https://github.com/Azure/service-fabric-issues/issues/1512), but my hopes are pretty low since nobody has even acknowledged the issue on the Microsoft side of things.

**Comment by JustinKaffenberger** [comment]
I've re-opened this issue as we are still seeing this issue on version `7.0.464.9590` of the Service Fabric runtime. The root of the issue has to do with `Get-ServiceFabricClusterConfiguration` returning a superfluous element that, when supplied to `Start-ServiceFabricClusterConfigurationUpgrade`, cause the exception:

> Authentication type cannot be changed from unsecured to Windows.

The superfluous element in question is called **WindowsIdentities** and can be seen in the following snippet (trimmed down for readability):

```
{
  "Properties": {
    "GoalStateExpirationReminderInDays": 0,
    "Security": {
      "$id": "1",
      "CertificateInformation": {
        "$id": "2",
        "ClusterCertificate": ...
        "ServerCertificate": ...
        "ReverseProxyCertificate": ...
        "ClientCertificateThumbprints": ...
      },
      "WindowsIdentities": {
        "$id": "3"
      },
      "ClusterCredentialType": "X509",
      "ServerCredentialType": "X509"
    },
    ...
    ...
}
```

As you can see, our desire is to have a pure X509 secured cluster with no Windows Identities, yet this element is being returned (empty no less). The existence of this element causes `Start-ServiceFabricClusterConfigurationUpgrade` to fail.

The workaround is to remove the **WindowsIdentities** prior to upgrading. This is troublesome since we are fond of automating a lot of our upgrade workflows, so we will have to build this workaround into our automation tooling.

**Comment by maburlik** [comment]
As mentioned in a few other thread on the same issue including https://github.com/microsoft/service-fabric-issues/issues/1512, the issue has been fixed as of 7.0 CU4 (7.0.470.9590).
